SingularityService/src/main/java/com/hubspot/singularity/SingularityLeaderController.java

```diff
@@ -71,6 +81,39 @@ public void start() {
 
   public void stop() {
     statePoller.finish();
+    TIMER.cancel();
+  }
+
+  @Override
+  public void stateChanged(CuratorFramework client, ConnectionState newState) {
+    LOG.info("Received update for new zk connection state {}", newState);
+    stateHandlerLock.lock();
+    try {
+      scheduler.setZkConnectionState(newState);
+      if (!newState.isConnected()) {
+        LOG.info("No longer connected to zk, pausing scheduler actions and waiting up to {}ms for reconnect",
+            configuration.getZooKeeperConfiguration().getAbortAfterConnectionLostForMillis());
+        lostConnectionStateChecker = new TimerTask() {
```

Comment:
The connection state can potentially go `SUSPENDED -> LOST -> RECONNECTED`, if the session times out. In this case, we would schedule a second `lostConnectionStateChecker`, which would override the first one. If we then reconnected, we would cancel the second lost state checker, but not the first. We probably want some sort of guard around this, either cancelling the existing task if it's present or not adding if there's a recent enough time task.